### ISSUE TITLE
Invoice Id is not stored into credit memo DB record

### ISSUE THREAD

**Issue by ytorbyk** [magento/magento2#6849]
### Preconditions
1. Magento 2.2.0-dev with sample data
2. Braintree payment method is configured and enabled
3. "Authorize" payment action is set for Braintree

### Steps to reproduce
1. Log in
2. Add at least two product into the shopping cart
3. Place order with Braintree payment method
4. Go to admin panel
5. Create two separate invoices (partial invoices) for the order
6. Open each of them and create credit memo for each one

### Expected result
1. Invoices ID is stored into DB record for the each credit memo in "sales_creditmemo" DB table

### Actual result
1. Invoice ID field is always empty for both cases - credit memo is crated for order or for invoice.

P.S. As improvement it would be nice to have an ability to show Invoice ID in credit memo grid additionally to Order ID.

**Comment by ytorbyk** [comment]
Here is pull request for the issue https://github.com/magento/magento2/pull/6902

**Comment by SerhiyShkolyarenko** [comment]
@ytorbyk Thank you for your contribution to Magento!
I performed all the steps you described, after step 5 i settled both transactions and continued step 6.
I checked sales_creditmemo table and found that invoice_id field contains correct invoice id.
Here is a data dump of my sales_creditmemo table:

```
INSERT INTO `sales_creditmemo` VALUES (1,1,0.0000,0.0000,0.0000,0.0000,1.0000,18.2000,0.0000,3.2000,15.0000,3.2000,0.0000,15.0000,0.0000,1.0000,0.0000,3.2000,0.0000,3.2000,0.0000,18.2000,0.0000,0.0000,0.0000,0.0000,1,NULL,NULL,NULL,2,1,2,1,'USD','USD','USD','USD','p068fdg0-refund','000000001','2016-10-20 15:00:53','2016-10-20 15:00:53',0.0000,0.0000,NULL,NULL,15.0000,15.0000,NULL,NULL,NULL),(2,1,0.0000,0.0000,0.0000,0.0000,1.0000,6.4000,0.0000,6.4000,0.0000,6.4000,0.0000,0.0000,0.0000,1.0000,0.0000,6.4000,0.0000,6.4000,0.0000,6.4000,0.0000,0.0000,0.0000,0.0000,1,NULL,NULL,NULL,2,1,2,2,'USD','USD','USD','USD','qqxfge0p-refund','000000002','2016-10-20 15:01:58','2016-10-20 15:01:58',0.0000,0.0000,NULL,NULL,0.0000,0.0000,NULL,NULL,NULL);
```

Am I doing anything wrong?

**Comment by ytorbyk** [comment]
@SerhiyShkolyarenko This is very weird. I retested it on the develop branch again and I have another result. 
<img width="913" alt="screen shot 2016-10-31 at 00 41 29" src="https://cloud.githubusercontent.com/assets/6418650/19840676/265cc5ae-9f04-11e6-96f4-1285d3a54ae6.png">

**Comment by SerhiyShkolyarenko** [comment]
@ytorbyk have you settled transaction after placing order?
Could you please put your credentials to this script and settle transaction with it?
[braintree_settle.zip](https://github.com/magento/magento2/files/561968/braintree_settle.zip)
Transaction id should be a parameter of the script.

**Comment by SerhiyShkolyarenko** [comment]
Offline refunds(credit memos) are not associated with invoices 1:1
It's possible to refund several invoices with one offline refund transaction, so in this case invoice_id will be NULL in sales_credit_memo record.

**Comment by SerhiyShkolyarenko** [comment]
@ytorbyk we analyzed our offline refunding again and concluded it needs improvement. After implementing story MAGETWO-60332 our credit memos will be bound strictly to invoices.
Thank you for taking part.

**Comment by piotrekkaminski** [comment]
Thank you for your submission.

We recently made some changes to the way we process GitHub submissions to more quickly identify and respond to core code issues.

Feature Requests and Improvements should now be submitted to the new [Magento 2 Feature Requests and Improvements](https://community.magento.com/t5/Magento-2-Feature-Requests-and/idb-p/feature-requests) forum (see details [here](https://community.magento.com/t5/News-Announcements/Improvements-to-GitHub-Management/m-p/44572#M96)).



We are closing this GitHub ticket and have moved your request to the new [forum](https://community.magento.com/t5/Magento-2-Feature-Requests-and/Invoice-Id-is-not-stored-into-credit-memo-DB-record/idi-p/54652).

**Comment by lazyguru** [comment]
@piotrekkaminski I don't see this as a feature request.  It is very much a bug.  Because invoice_id is not set, the relationship of the credit memo to the invoice is not persisted.  This affects ONLINE refunds as much as offline ones.  

**Scenario**
(tested on 2.1.2)
* Purchase ItemA with qty 4
* Invoice ItemA on InvoiceA for qty 2
* Invoice ItemA on InvoiceB for qty 1
* CreditMemo against InvoiceA for qty 1 (invoice_id is not persisted in sales_creditmemo table)
* Create a new CreditMemo against InvoiceA  <-- Here the qty field is defaulted to "2" even though 1 unit was previously refunded already.

The impact of this is for external payment processors that tie refunds to specific invoices and expect quantities to match.  In the scenario above, Magento will have refunded 3 units of ItemA against InvoiceA, however the payment processor may only credit the customer for the refund of 2 units (since that is all that was captured on InvoiceA).

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-69915

**Comment by lazyguru** [comment]
@magento-team  Thanks for the internal ticket update.  Can you also reopen this issue so it is clear that it is being tracked?

**Comment by ytorbyk** [comment]
Thank you!
Can you please reopen this ticker as well?
